---
title: "HER2_Amp_Patients"
author: "Aurora S Blucher"
date: "3/4/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(here)
library(readxl)
```

## R Markdown

Script for TCGA patients; HER2

Notes 03/04/21
-TCGA breast patients, HER2 by IHC subset
-pull in TCGA clinical/survival data (using clinical data resource*)
-pull in CN and expression
-compare between the 2 groups

```{r}
#tcga patients
#clinical, get all BRCA patients
TCGASamplesAll<-read_excel(here("data", "TCGA_PANCAN_fromXena", "TCGA-CDR-SupplementalTableS1.xlsx"))
#View(TCGASamplesAll)

BRCASamples_Clinical<-TCGASamplesAll %>%
  filter(type=="BRCA")%>%
  select(Sample_ID = bcr_patient_barcode, type, histological_type, OS, OS.time)
#View(BRCASamples_Clinical) #1097 total

#TCGA COPY NUMBER info
#erbb2, subset to brca
#then require 2 or 1 (amplified)
#
TCGA_GISTIC<-read_tsv(here("data", "TCGA_PANCAN_fromXena", "TCGA.PANCAN.sampleMap_Gistic2_CopyNumber_Gistic2_all_thresholded.by_genes"))
head(TCGA_GISTIC)

TCGA_ERBB2<-TCGA_GISTIC %>%
  filter(Sample=="ERBB2")%>% # a little funny because we need genes but this first column is titled 'Sample'
  pivot_longer(-Sample, names_to = "Sample_ID", values_to = "GISTIC_Score") %>%
  mutate(Sample_ID_Abbr = str_sub(Sample_ID, 0, -4L)) %>%
  select(Sample_ID_Full=Sample_ID, Sample_ID=Sample_ID_Abbr, GeneName = Sample, GISTIC_Score)  
View(TCGA_ERBB2)
#output to file
write_tsv(TCGA_ERBB2, here("data_intermediate", "TCGA_PanCan_fromXena_GISTIC2_ERBB2.tsv"))

#okay now join with our TCGA BRCA subset above
BRCASamples_Clinical_joinERBB2<-left_join(BRCASamples_Clinical, TCGA_ERBB2)
View(BRCASamples_Clinical_joinERBB2)
#output to file
write_tsv(BRCASamples_Clinical_joinERBB2, here("data_intermediate", "TCGA_PanCan_fromXena_BRCA_ERBB2.tsv"))

#now let's look at the distribution of survival times here
#long tail out to right, Samuel wants to use 3 years as our cutoff
hist_survival<-BRCASamples_Clinical_joinERBB2 %>%
  filter(GISTIC_Score==2) %>%
  ggplot(aes(x=OS.time))+
  geom_histogram(color = "black", fill = "gray75")+
  geom_vline(aes(xintercept=1095),
            color="blue", linetype="dashed", size=1)+
  annotate(geom="text", x=1650, y=25, label="3YR Survival",
              color="blue")+
  labs(title="Distribution of OS in TCGA BRCA, ERBB2 Amplified Patients", subtitle="Requiring GISTIC Score==2")+
  xlab("Overall Survival (days)")+
  ylab("Count")
hist_survival

HER2AmpSamples<-BRCASamples_Clinical_joinERBB2 %>%
  filter(GISTIC_Score==2)
View(HER2AmpSamples)#133 total, write to file
write_tsv(HER2AmpSamples, here("data_intermediate", "TCGA_PanCan_fromXena_BRCA_ERBB2_Amplified.tsv"))

HER2AmpSamples<-read_tsv(here("data_intermediate", "TCGA_PanCan_fromXena_BRCA_ERBB2_Amplified.tsv"))

#Under 3 years
HER2Amp_Under3Years<-HER2AmpSamples%>%
  filter(GISTIC_Score==2) %>%
  filter(OS.time<1095)
View(HER2Amp_Under3Years) #79 samples

#At or over 3 years
HER2Amp_Over3Years<-HER2AmpSamples%>%
  filter(GISTIC_Score==2) %>%
  filter(OS.time>=1095)
View(HER2Amp_Over3Years) #54 samples

```

```{r}
#okay get all the gistic data for these 133 samples
head(TCGA_GISTIC)
TCGA_GISTIC_HER2Amp_Samples<-TCGA_GISTIC %>%
  select(Sample, one_of(HER2AmpSamples$Sample_ID_Full)) #not working?
View(TCGA_GISTIC_HER2Amp_Samples)
#write to file
write_tsv(TCGA_GISTIC_HER2Amp_Samples, here("data_intermediate", "TCGA_PanCan_fromXena_GISTIC_HER2AmpSamples.tsv"))

#go long, so we can annotate
#then filter to just our +2 calls, b/c we only want amplifications
TCGA_GISTIC_HER2Amp_SamplesLONG<-TCGA_GISTIC_HER2Amp_Samples %>%
  pivot_longer(-Sample, names_to = "Sample_ID", values_to = "GISTIC_Score") %>% #ugh the sample ID here should be 'gene'
  mutate(Group = ifelse(Sample_ID %in% HER2Amp_Under3Years$Sample_ID_Full, "Short_Under3Years", "Long_Over3Years")) %>%
  select(Sample_ID_Full = Sample_ID, Group, Gene=Sample, GISTIC_Score) %>%
  filter(GISTIC_Score==2)

View(TCGA_GISTIC_HER2Amp_SamplesLONG)

#now create grouped data frame and count up per gene; how many in short versus long groups?
TCGA_GISTIC_HER2Amp_SamplesLONG_GroupCounts<-TCGA_GISTIC_HER2Amp_SamplesLONG %>%
  group_by(Gene, Group) %>%
  tally()
View(TCGA_GISTIC_HER2Amp_SamplesLONG_GroupCounts)
  
#back to wide
TCGA_GISTIC_HER2Amp_SamplesLONG_GroupCounts_WIDE<-TCGA_GISTIC_HER2Amp_SamplesLONG_GroupCounts %>%
  pivot_wider(names_from = Group, values_from= n, values_fill=0 ) #fill in our NAs with a 0 count
View(TCGA_GISTIC_HER2Amp_SamplesLONG_GroupCounts_WIDE)
#output filee
write_tsv(TCGA_GISTIC_HER2Amp_SamplesLONG_GroupCounts_WIDE, here("data_intermediate", "TCGA_PanCan_fromXena_GISTIC_HER2AmpSamples_SurvivalGroupCounts.tsv"))

#rename date farme
#dataFrame_GroupCounts<-TCGA_GISTIC_HER2Amp_SamplesLONG_GroupCounts_WIDE %>%
#  pivot_longer(-Gene)
#View(dataFrame_GroupCounts)

#set up with broom so we can iterate and run proportion tests
library(broom)
#prop.test set up
x<-c(12, 26) #cases
n<-c(24, 28) #counts of the trial
prop.test(x, n) #compare counts versus

#start with a small sample to get broom set up first
View(TCGA_GISTIC_HER2Amp_SamplesLONG_GroupCounts_WIDE[1:10,]) #just 10 here

#let's see if this works; otherwise put it away 7:30pm
#prop.test might need table() command, or 2 vectors, not sure
#check to make sure these numbers are right based on example
dataFrame_PropTest<-TCGA_GISTIC_HER2Amp_SamplesLONG_GroupCounts_WIDE[20:50,]%>% 
  group_by(Gene) %>%
  mutate(test = list(broom::tidy(prop.test(c(Long_Over3Years, Short_Under3Years), c(54,79), 
                                           conf.level = .95, alternative="less")))) %>%
  tidyr::unnest(test)
View(dataFrame_PropTest)

#check 12 long, 17 short | yes this matches our data table above, first entry; good
x<-c(2, 0) #cases
n<-c(54, 79) #counts of the trial
prop.test(x, n, conf.level = .95) #compare counts versus


#see example from here:


#now, we only want what is amplified in SHORT versus LONG
#run 7:19PM -> how long?
#alternative should be LESS, since we want to test that LONG< SHORT
#i.e. that the count of amplifications is LESS IN LONG SURVIVAL versus SHORT SURVIVAL
dataFrame_PropTestFull<-TCGA_GISTIC_HER2Amp_SamplesLONG_GroupCounts_WIDE%>% 
  filter(Short_Under3Years>0) %>% #ADD FILTER HERE
  group_by(Gene) %>%
  mutate(test = list(broom::tidy(prop.test(c(Long_Over3Years, Short_Under3Years), c(54,79), 
                                           alternative = "less", conf.level = .95)))) %>%
  tidyr::unnest(test)
View(dataFrame_PropTestFull)
#write to file
write_tsv(dataFrame_PropTestFull, here("output", "PropTest_LongGroupLessAmpThanShortGroupAmp_Results_030421.tsv"))

```

```{r}
#start rna-seq section
#our 133 erbb2 samples
HER2Samples_OurSet<-read_tsv(here("data_intermediate", "TCGA_PanCan_fromXena_BRCA_ERBB2_Amplified.tsv"))
View(HER2Samples_OurSet)

#okay now try adapting the tcgabiolinks vignette for our samples
library(TCGAbiolinks)

#we don't have the full barcodes, so this may not work
listSamples <- HER2Samples_OurSet$Sample_ID_Full

# Query platform Illumina HiSeq with a list of barcode #?Working so far for our BRCA samples
query <- GDCquery(project = "TCGA-BRCA", 
                  data.category = "Gene expression",
                  data.type = "Gene expression quantification",
                  experimental.strategy = "RNA-Seq",
                  platform = "Illumina HiSeq",
                  file.type = "results",
                  barcode = listSamples, 
                  legacy = TRUE)
class(query)
View(query)
class(query$barcode)
View(query$barcode)

# Download a list of barcodes with platform IlluminaHiSeq_RNASeqV2
GDCdownload(query)

# Prepare expression matrix with geneID in the rows and samples (barcode) in the columns
# rsem.genes.results as values
BRCARnaseqSE <- GDCprepare(query)
View(BRCARnaseqSE)

library(SummarizedExperiment)
BRCAMatrix <- assay(BRCARnaseqSE,"raw_count") # or BRCAMatrix <- assay(BRCARnaseqSE,"raw_count")
View(BRCAMatrix) #132 samples X 19,947 genes

####skip

# REVIEW this step
# For gene expression if you need to see a boxplot correlation and AAIC plot to define outliers you can run
BRCARnaseq_CorOutliers <- TCGAanalyze_Preprocessing(BRCARnaseqSE)
BRCARnaseq_CorOutliers

library(EDASeq)
library(edgeR)

# normalization of genes
dataNorm <- TCGAanalyze_Normalization(tabDF = dataBRCA, geneInfo =  geneInfo)

# quantile filter of genes
dataFilt <- TCGAanalyze_Filtering(tabDF = dataNorm,
                                  method = "quantile", 
                                  qnt.cut =  0.25)
View(dataFilt)

#now we need to assign SHORT vs LONG groups

#######

View(BRCAMatrix)
#get all the column names = samlple names as queries
BRCAMatrix_sampleNamesFull<-colnames(BRCAMatrix)
View(BRCAMatrix_sampleNamesFull)
BRCAMatrix_sampleNamesFull<-as.data.frame(BRCAMatrix_sampleNamesFull) #data frame first so we can set colnames
colnames(BRCAMatrix_sampleNamesFull)<-c("Sample_ID_FullBarcode") 

BRCAMatrix_sampleNamesFull_Abbr<-BRCAMatrix_sampleNamesFull %>%
  mutate(Sample_ID = str_sub(Sample_ID_FullBarcode, 0, 12))
View(BRCAMatrix_sampleNamesFull_Abbr)

#now join with our sample information
BRCAMatrix_sampleNamesFull_Abbr_toGroup<-BRCAMatrix_sampleNamesFull_Abbr %>%
  mutate(Group = ifelse(Sample_ID %in% HER2Amp_Over3Years$Sample_ID, "Long", "Short"))
View(BRCAMatrix_sampleNamesFull_Abbr_toGroup)


#WORKING HERE

#okay, now need to split according to this indicator
samplesSHORT<-BRCAMatrix_sampleNamesFull_Abbr_toGroup %>%
  filter(Group=="Short") %>%
  select(Sample_ID_FullBarcode)
View(samplesSHORT) #78 entries, matches

# selection of tumor samples "TP"
samplesLONG<-BRCAMatrix_sampleNamesFull_Abbr_toGroup %>%
  filter(Group=="Long") %>%
  select(Sample_ID_FullBarcode)
View(samplesLONG) #54 entries, matches

#now we should be able to subset according to these
BRCADataShort<-as.data.frame(BRCAMatrix) %>%
  select(one_of(samplesSHORT$Sample_ID_FullBarcode))
View(BRCADataShort)
BRCAMatrixShort<-as.matrix(BRCADataShort)
View(BRCAMatrixShort)

BRCADataLong<-as.data.frame(BRCAMatrix) %>%
  select(one_of(samplesLONG$Sample_ID_FullBarcode))
View(BRCADataLong)
BRCAMatrixLong<-as.matrix(BRCADataLong)



# Diff.expr.analysis (DEA)
dataDEGs <- TCGAanalyze_DEA(mat1 = dataFilt[,samplesNT],
                            mat2 = dataFilt[,samplesTP],
                            Cond1type = "Normal",
                            Cond2type = "Tumor",
                            fdr.cut = 0.01 ,
                            logFC.cut = 1,
                            method = "glmLRT")




#if this runs, need to pull boxplots to check

```

